### PR TITLE
feat(tests): implement base tests for providers

### PR DESCRIPTION
**Pull request by WilsontheWolf**
Makes all tests share one codebase.
Might want to fix some stuff up I just kinda @ts-ignored everything.
Also commented out Provider.Identifiers lines until RealShadowNova/joshdb-core#7 is merged.

### CODE REVIEW

File: tsconfig.eslint.json [hunk 1]
```diff
@@ -7,6 +7,7 @@
 		"packages/**/jest.config.ts",
 		"packages/**/scripts/rollup.config.ts",
 		"packages/**/tests/*.ts",
-		"scripts"
+		"scripts",
+		"tests/base.ts"
```
Comment (RealShadowNova) [inline tsconfig.eslint.json:11]:
```suggestion
		"tests/**/*.ts
```

File: tests/base.ts [hunk 2]
```diff
@@ -0,0 +1,1765 @@
+import { JoshProvider, JoshProviderError, MathOperator, Method, Payload } from '@joshdb/core';
+
+export default (Provider: typeof JoshProvider, ProviderError: typeof JoshProviderError, providerData: unknown) => {
+	describe('Provider', () => {
```
Comment (RealShadowNova) [inline tests/base.ts:4]:
```suggestion
	describe(Provider.prototype.constructor.name, () => {
```

File: tests/base.ts [hunk 3]
```diff
@@ -0,0 +1,1765 @@
...
+		describe('is a class', () => {
+			test('GIVEN typeof Provider THEN returns function', () => {
```
Comment (RealShadowNova) [inline tests/base.ts:6]:
```suggestion
			test(`GIVEN typeof ${Provider.prototype.constructor.name} THEN returns function`, () => {
```

File: tests/base.ts [hunk 4]
```diff
@@ -0,0 +1,1765 @@
+import { JoshProvider, JoshProviderError, MathOperator, Method, Payload } from '@joshdb/core';
+
+export default (Provider: typeof JoshProvider, ProviderError: typeof JoshProviderError, providerData: unknown) => {
```
Comment (RealShadowNova) [inline tests/base.ts:3]:
```suggestion
import { JoshProvider, JoshProviderError, MathOperator, Method, Payload } from '@joshdb/core';

export default (Provider: typeof JoshProvider, ProviderError: typeof JoshProviderError, providerOptions: JoshProvider.Options) {
```

**Review by RealShadowNova — CHANGES_REQUESTED** [review]
It's a good start, please rename the `tests/base.ts` file to the name of the function being exported.

File: packages/json/tests/lib/JSONProvider.test.ts [hunk 5]
```diff
@@ -1,1678 +1,4 @@
-import { MathOperator, Method, Payload } from '@joshdb/core';
+import tests from '../../../../tests/base';
```
Comment (RealShadowNova) [inline packages/json/tests/lib/JSONProvider.test.ts:1]:
```suggestion
import { runProviderTest } from '../../../../tests/runProviderTest';
```

File: packages/json/tests/lib/JSONProvider.test.ts [hunk 6]
```diff
@@ -1,1678 +1,4 @@
...
-
-				expect(method).toBe(Method.Values);
-				expect(trigger).toBeUndefined();
-				expect(error).toBeUndefined();
-				expect(data).toEqual(['value']);
-			});
-		});
-	});
-});
+tests(JSONProvider, JSONProviderError, { dataDirectoryName: '.tests' });
```
Comment (RealShadowNova) [inline packages/json/tests/lib/JSONProvider.test.ts:4]:
```suggestion
runProviderTests<JSONProvider.Options>(JSONProvider, JSONProviderError, { dataDirectoryName: '.tests' });
```

File: packages/mongo/tests/lib/MongoProvider.test.ts [hunk 7]
```diff
@@ -1,1729 +1,5 @@
-import { MathOperator, Method, Payload } from '@joshdb/core';
+import tests from '../../../../tests/base';
```
Comment (RealShadowNova) [inline packages/mongo/tests/lib/MongoProvider.test.ts:1]:
```suggestion
import { runProviderTest } from '../../../../tests/runProviderTest';
```

File: packages/mongo/tests/lib/MongoProvider.test.ts [hunk 8]
```diff
@@ -1,1729 +1,5 @@
...
-
-				expect(method).toBe(Method.Values);
-				expect(trigger).toBeUndefined();
-				expect(error).toBeUndefined();
-				expect(data).toEqual(['value']);
-			});
-		});
-	});
-});
+tests(MongoProvider, MongoProviderError, { collectionName: 'josh-test' });
```
Comment (RealShadowNova) [inline packages/mongo/tests/lib/MongoProvider.test.ts:5]:
```suggestion
runProviderTest<MongoProvider.Options>(MongoProvider, MongoProviderError, { collectionName: 'josh-test' });
```

**Review by RealShadowNova — CHANGES_REQUESTED** [review]
Just some more changes rq

**Comment by WilsontheWolf** [comment]
> please rename the `tests/base.ts` file to the name of the function being exported

Would that be `tests/tests.ts` or `tests/index.ts`

**Comment by RealShadowNova** [comment]
> > please rename the `tests/base.ts` file to the name of the function being exported
> 
> Would that be `tests/tests.ts` or `tests/index.ts`

It would be like `tests/runProviderTest.ts`

File: tests/runProviderTest.ts [hunk 9]
```diff
@@ -1,9 +1,13 @@
 import { JoshProvider, JoshProviderError, MathOperator, Method, Payload } from '@joshdb/core';
 
-export default (Provider: typeof JoshProvider, ProviderError: typeof JoshProviderError, providerData: unknown) => {
-	describe('Provider', () => {
+export function runProviderTest<ProviderOptions extends JoshProvider.Options>(
+	Provider: typeof JoshProvider,
+	ProviderError: typeof JoshProviderError,
+	providerOptions: ProviderOptions
+) {
```
Comment (RealShadowNova) [inline tests/runProviderTest.ts:7]: This should hopefully fix all of those `@ts-expect-error`s.

```suggestion
export function runProviderTest<Provider extends JoshProvider, ProviderOptions extends JoshProvider.Options>(
	Provider: typeof Provider,
	ProviderError: typeof JoshProviderError,
	providerOptions: ProviderOptions
) {
```

File: tests/runProviderTest.ts [hunk 10]
```diff
@@ -14,10 +18,10 @@ export default (Provider: typeof JoshProvider, ProviderError: typeof JoshProvide
 
 		describe('can manipulate provider data', () => {
 			// @ts-expect-error Cause I can.
-			const provider = new Provider(providerData);
+			const provider = new Provider(providerOptions);
 
 			beforeAll(async () => {
-				await provider.init({ name: 'joshtest' });
+				await provider.init({ name: 'josh-test' });
```
Comment (RealShadowNova) [inline tests/runProviderTest.ts:24]:
```suggestion
				await provider.init({ name: 'provider' });
```

**Review by RealShadowNova — CHANGES_REQUESTED** [review]
If it becomes necessary, which I don't believe it will be, you can add a generic for `ProviderError`. Other then that, it looks good at this time.

File: packages/mongo/tests/lib/MongoProvider.test.ts [hunk 11]
```diff
@@ -1,1729 +1,5 @@
...
-
-				expect(method).toBe(Method.Values);
-				expect(trigger).toBeUndefined();
-				expect(error).toBeUndefined();
-				expect(data).toEqual(['value']);
-			});
-		});
-	});
-});
+runProviderTest<MongoProvider.Options>(MongoProvider, MongoProviderError, { collectionName: 'josh-test' });
```
Comment (RealShadowNova) [inline packages/mongo/tests/lib/MongoProvider.test.ts:5]: `collectionName` falls back to the name passed in `init()` so this isn't needed.

```suggestion
runProviderTest<MongoProvider.Options>(MongoProvider, MongoProviderError);
```

**Review by RealShadowNova — CHANGES_REQUESTED** [review]

File: packages/json/tests/lib/JSONProvider.test.ts [hunk 12]
```diff
@@ -1,1678 +1,4 @@
...
-
-				expect(method).toBe(Method.Values);
-				expect(trigger).toBeUndefined();
-				expect(error).toBeUndefined();
-				expect(data).toEqual(['value']);
-			});
-		});
-	});
-});
+runProviderTest<JSONProvider.Options>(JSONProvider, JSONProviderError, { dataDirectoryName: '.tests' });
```
Comment (RealShadowNova) [inline packages/json/tests/lib/JSONProvider.test.ts:4]:
```suggestion
runProviderTest<JSONProvider, JSONProvider.Options>(JSONProvider, JSONProviderError, { dataDirectoryName: '.tests' });
```

File: packages/mongo/tests/lib/MongoProvider.test.ts [hunk 13]
```diff
@@ -1,1729 +1,5 @@
...
-
-				expect(method).toBe(Method.Values);
-				expect(trigger).toBeUndefined();
-				expect(error).toBeUndefined();
-				expect(data).toEqual(['value']);
-			});
-		});
-	});
-});
+runProviderTest<MongoProvider.Options>(MongoProvider, MongoProviderError);
```
Comment (RealShadowNova) [inline packages/mongo/tests/lib/MongoProvider.test.ts:5]:
```suggestion
runProviderTest<MongoProvider, MongoProvider.Options>(MongoProvider, MongoProviderError);
```

**Review by RealShadowNova — CHANGES_REQUESTED** [review]

**Review by RealShadowNova — APPROVED** [review]

**Pull request merged**